### ISSUE TITLE
Cron manager

### ISSUE THREAD

**Issue by anddimario** [githubsaturn/captainduckduck#38]
Hi, great project, what do you think about the possibility to manage cron?
For example, on master, something like app manager, where we can define the instances/server, the schedule and upload a script. The processes' balance could be a simple round robin, or something based on instance load. The master instance run the cron based on schedule and execute the process on the right instance.

**Comment by githubsaturn** [comment]
You can do all the things that you explain in one container or one master and multiple worker container. As far as Captain is concerned, once you deploy your containers, you can do whatever you want with them.

To provide additional level of customization, you can invoke methods on captain through its API. I'll be writing some documentation on how to use captain api soon. Stay tuned.

But in general, keep in mind tgat Captain is a deployment platform not a process manager. So, you can use your favorite cron manager in a container and it won't interfere with Captain at all.

**Comment by avidale** [comment]
Hi everyone!
I am wondering, whether there is a simple solution that can schedule regular jobs within Captain environment. 
Maybe, you can hint at one that you recommend.

**Comment by githubsaturn** [comment]
There is no preferred way. You can simply create a container that runs a cron job. Read more:
https://stackoverflow.com/questions/37458287/how-to-run-a-cron-job-inside-a-docker-container

**Comment by maietta** [comment]
> There is no preferred way. You can simply create a container that runs a cron job. Read more: https://stackoverflow.com/questions/37458287/how-to-run-a-cron-job-inside-a-docker-container

I have discovered a project at https://github.com/mcuadros/ofelia which seems to be a nice option for those who deploy their applications via the Docker Compose method standalone or with the use of Caprover.

Fellow dev here, looking at ways I can help the Caprover project along. Being that Caprover is a deployment management tool and is completely optional, I have grappled with the idea of implementing a crontab scheduler in the API and frontend UI, or just doing a write-up instead on how to implement something like https://github.com/mcuadros/ofelia in Caprover for others who are also looking for a solution.

I don't want to see Caprover become bloated but I can see the reasoning why this would be a desired feature. I can envision in my head a new Apps "tab" called "Scheduler" with an easy to understand interface. Perhaps the implementation of a "Enable Ofelia Scheduler" option much like the Netdata utility would be the way to go about it. Then that would enable the "Scheduler" tab for the Apps. The API can be extended to enable the management of the scheduler for apps.

I would like to know other people's thoughts on this, especially from the people who regularly contribute to Caprover. (Doesn't seem like that many, sadly).

**Comment by githubsaturn** [comment]
> I don't want to see Caprover become bloated...

This is exactly why we have this [section here](https://github.com/caprover/caprover/blob/master/CONTRIBUTING.md#important-caprover-goals--scope).


The long-term goal is to keep CapRover's included batteries as thin as possible and allow optional extensions. NetData is an interesting one. I'm considering moving it to One Click apps. This requires more investment on one click apps to allow customization. But that's the effort that is definitely worth taking on. 

So for job scheduler, consider the possibility of being able to deploy a "Scheduler" as a one click app. 

Looking at Ofelia, it's super simple to deploy it as a one click app except the fact that it requires customizations on labels and etc. Those can simply be addressed by the [Service Override object](https://caprover.com/docs/service-update-override.html).

So, to sum up, you can build Ofelia as a one click app. Deploy the instance. Then to add jobs, you'll have to add labels to service override box, e.g.:
```
Labels:
   - key1 : val1
   - key2 : val2
```

**Comment by maietta** [comment]
Thanks for the feedback. It had not occurred to me to prepare a one-click-app for this. I've got some free time this week so I'll see what I can put together and if it works out well, I'll prepare a pull request per the guide. I am not totally familiar with how to approach the creation and testing of a One Click App for Caprover, so it could take me a bunch of trial and error to get it right.

**Comment by githubsaturn** [comment]
Yea probably something like this should work:
```yaml
captainVersion: 4
services:
    $$cap_appname:
    image: 
    dockerfileLines:
        - FROM mcuadros/ofelia:cap_version
        - CMD daemon --docker
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock:ro
    caproverExtra:
        notExposeAsWebApp: 'true'
caproverOneClickApp:
    variables:
        - id: $$cap_version
          label:  Version
          defaultValue: latest
```

And then post deployment, add this bit to your service override:
```yaml
Labels:
   - ofelia.job-local.my-test-job.schedule :  "@every 5s"
   - ofelia.job-local.my-test-job.command : "date"

```

**Comment by maietta** [comment]
Well, I seem to have run into some trouble getting going on this.

I have successfully "validated" and "built" the one click app using NPM and now have a dist/ directory with just 1 app, the one I am working on. I have setup a Github action to package up and deploy the dist/ contents for deployment to Caprover using some CI/CD magic.

I have the one-click-app hostname configured at my Caprover instance at https://oneclickapps.premoweb.net, however Caprover won't let me connect up the url.

I've tried to add my 3rd party repo like so:
![image](https://user-images.githubusercontent.com/11745020/135491070-5d8af984-e0ca-47fc-98b6-75e740abf293.png)

But the error I get back is:
"1000 : Could not fetch app lists from oneclickapps.premoweb.net"

So my question now is so I can begin testing, do I need to point the web server to use dist/ or dist/v4/ for the public web root?

-- Stumped.

**Comment by maietta** [comment]
Hey @githubsaturn, thank you for the example. I did manage to get a One Click App configured up and working, it deploys fine but it wont' stay running as it's looking for an INI file. I am hitting a wall where I am not sure how to include a file with a one click app, so I'll have to go digging around for examples on how that can be added.

Additionally, I managed to overlook the simple concept of Ofelia that requires Service Labels, not Enviornment variables to let Ofelia know when and how to run a task.

There are two solutions I can go with this if I wish to implement Ofelia sheduler system to work nice with Caprover:

1) Fork Ofelia as a new project, rewriting it to use environment variables key-pairs rather than labels or
2) or Request a feature for Caprover to let us manage labels on services. (Which seems a bit overkill, but service label management does seems to make sense, at least in this case)

I am more inclined to just fork the Ofelia project and pair it back along with making a few changes to make it compatible with Caprover's UI by using Enviornment key-pairs rather than Service labels.

... now back to thinking mode ... and requesting yours and other's input on this topic.

Thank you.

**Comment by maietta** [comment]
Adding labels via Service Update Override section does not actually seem to apply labels. I inspect the service and the labels don't appear.

This is what I used. I couldn't get it to accept this without TaskTemplate: at the top.

TaskTemplate:
  Labels:
    - ofelia.enabled=true
    - ofelia.job-exec.quotefall-rotate.command: "curl https://quotefall.com/HELLO_FROM_CRONJOB"
    - ofelia.job-exec.quotefall-rotate.schedule: "@every 5s"

**Comment by githubsaturn** [comment]
My bad... Turned out it's not an array, it's a key-map, also it resides under container... Try this:

```yaml
TaskTemplate:
  ContainerSpec:
    Labels:
      some.label: some.value
```

**Comment by maietta** [comment]
Thanks! That totally worked.

So now I am in the final home stretch to get something ready for testing, then included as a one-click-app.

I couldn't get the one-click-app to work because it's complaining about a missing INI file, but I am able to run the service via Docker compose with no issues using the following in docker-compose.yml

`version: "3"
services:
  ofelia:
    image: mcuadros/ofelia:latest
    command: daemon --docker
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock:ro
    labels:
      ofelia.job-local.test.schedule: "@every 5s"
      ofelia.job-local.test.command: "date"`

Then in my App's Service Update Override section, I dropped in the following:

`TaskTemplate:
  ContainerSpec:
    Labels:
      ofelia.enabled: "true"
      ofelia.job-exec.quotefall-rotate.command: "curl https://quotefall.com/test"
      ofelia.job-exec.quotefall-rotate.schedule: "@every 3s"`

This totally works as expected.

I do have a one-click-app hostname setup for testing at https://oneclickapps.premoweb.net and have Apache exposing the contents of the directory so it's browsable. (Feel free to connect the app and deploy). The Github project repo is at https://github.com/PremoWeb/one-click-apps. (This is for my my testing, once finished, I will do a pull request from another fork that only focuses on what's needed to include the app)

So I am down to 2 issues:

- How to get the Ofelia service to recognize when a container is updated, without having to restart the Ofelia service.
- How to get the One-Click-App to work correctly and not complain about a missing INI.

I can almost smell a fresh new One Click App baking. Any further input is appreciated.

**Comment by githubsaturn** [comment]
It's most likely due to the shell format of CMD, try changing it to this:
```
        dockerfileLines:
          - FROM mcuadros/ofelia:$$cap_version
          - CMD ["daemon", "--docker"]
```

**Comment by githubsaturn** [comment]
Also, make sure you remove this line from your one click app: `image: mcuadros/ofelia:latest` - it overrides dockerfileLines

**Comment by maietta** [comment]
> It's most likely due to the shell format of CMD, try changing it to this:
> 
> ```
>         dockerfileLines:
>           - FROM mcuadros/ofelia:$$cap_version
>           - CMD ["daemon", "--docker"]
> ```

This has been completed yet it's still kicking my butt. 

Service finally showing present, but not staying running:
lep4ki74wvzf   srv-captain--a                            replicated   0/1        mcuadros/ofelia:latest

Strangely enough, no error logs. In fact, no logs at all for this service.

Hmm...

**Comment by githubsaturn** [comment]
If you see this: `lep4ki74wvzf srv-captain--a replicated 0/1 mcuadros/ofelia:latest` then it means you probably didn't remove the `image` from your one-click app definition. It's not using the dockerfileLines. 

I expect you to see something like (img-captain-a:1)
```
lep4ki74wvzf srv-captain--a replicated 0/1 img-captain-a:1
```

**Comment by githubsaturn** [comment]
Use this:
```yaml
captainVersion: 4
services:
    '$$cap_appname':
        dockerfileLines:
          - FROM mcuadros/ofelia:$$cap_version
          - CMD ["daemon", "--docker"]
        volumes:
            - /var/run/docker.sock:/var/run/docker.sock:ro
        caproverExtra:
            notExposeAsWebApp: 'true'
caproverOneClickApp:
    variables:
        - id: '$$cap_version'
          label: Ofelia version
          defaultValue: 'latest'
          description: A cron scheduler
          validRegex: "/^([^\\s^\\/])+$/"
        - id: '$$cap_tz'
          label: Time Zone
          defaultValue: America/New_York
          description: Get yours from https://en.wikipedia.org/wiki/List_of_tz_database_time_zones
          validRegex: '/.{1,}/'
    instructions:
        start: |-
            A nice cron utility 
        end: |-
            Ofelia has been successfully deployed!
    displayName: Ofelia
    isOfficial: false
    description: A cron utility
    documentation: No docs yet
```

**Comment by maietta** [comment]
So I now keep getting this error on my caprover-definition, but like this: This is for my self-hosted one-click-app endpoint. This is were I am pulling the app together.

```
  {
  "schemaVersion": 2,
  "dockerfileLines": [
        "FROM httpd:2.4",
        "RUN rm -r /usr/local/apache2/htdocs/*",
        "COPY ./dist/ /usr/local/apache2/htdocs/",
        "COPY ./.htaccess /user/local/apache2/htdocs/"
    ]
 }
```
![image](https://user-images.githubusercontent.com/11745020/135559556-00a13e63-7db3-4fc5-999a-ae935905ef9c.png)

I've doing the best I can to get this nailed down tonight. It's been a long day trying to figure all this out. Pretty nice system, overall. Just a bit of a learning curve is all.

Feels like Caprover *might* be caching the app that is fetched?

**Comment by githubsaturn** [comment]
I'm not sure what you're trying to do at this point. What's the `captain-definition` for? For one click app, all you need is the yaml file that I pasted above. Why do you need this captain-definition file?

**Comment by maietta** [comment]
> I'm not sure what you're trying to do at this point. What's the `captain-definition` for? For one click app, all you need is the yaml file that I pasted above. Why do you need this captain-definition file?

This file was added so that I can automate the rebuild and deployment of my "for testing" own One Click App repository:

You see, I couldn't figure out how to deploy a one-click-app I was working on, using the ofelia.yml I was working on. I could only see that I can deploy  via a packaged tar file, a direct docker image, or a captain-definition file containing everything, which I don't yet have prepared for Ofelia.

To answer your question the **captain-definition**, **Dockerfile** and the .github/workflows/**deploy-production.yml** files were added so that I can streamline the automatic re-deployments of my "for testing" One Click App repository that contains just 1 ofelia.yml project. After using the included validate and build functions, I can then commit and push my changes and within a minute or so, Caprover has a new instance of my One Click App repository deployed. To make sure I can browse around this repository at the https://oneclickapps.premoweb.net/ web address, I've also slipstreamed in a .htaccess file as the deployment of that is using Apache and this is how I'm telling Apache that I want to see the contents of the directory, for inspection.

**Comment by githubsaturn** [comment]
Ah I see. So at this point, the problem that you're having is with the deployment of the one click app repository. I'll look into that this afternoon, but it should be fairly simple.

PS1: as I mentioned above, for testing, you don't nee to deploy your own repository. You can just copy paste the YAML definition on CapRover's control panel.

PS2: I personally deploy the one click app repository on Github pages, they have a much better latency for the official one click app. But definitely you should also be able to deploy it on CapRover. I'll review it tonight and figure out what the issue is.

**Comment by maietta** [comment]
Thanks for the quick reply. I know you've put a lot into Caprover and I very much appreciate your assistance.

When I got started working on my first One Click App, I couldn't get the CNAME publishing to work in Github pages. For whatever reason i've never had trouble before. So I just opted to publish on my own server, leading me to this.

Regarding PS1: If I can deploy the YAML file directly in the Caprover UI, then where is it? There are a number of methods for deploying to Caprover, but none of which seem to use YAML. Unless I am missing something here.....

**Comment by maietta** [comment]
UPDATE: Got the app deployed now!! 

My issue was related to nesting. Now it makes sense.. as it seems that Caprover converts the YAML into a caprover-definition format prior to processing for deployment. As such, it really couldn't find the dockerfileLines because that was not being parsed as something caprover was looking for, because it was outside the scope. Once the dockerfileLines section was moved to the caproverExtra block, now Caprover can see it! I'm learning!

INCORRECT (not working, kept complaining):
```
captainVersion: 4
version: 3.3
services:
    '$$cap_appname':
        dockerfileLines:
          - FROM mcuadros/ofelia:$$cap_version
          - CMD ["daemon", "--docker"]
        volumes:
            - /var/run/docker.sock:/var/run/docker.sock:ro
        caproverExtra:
            notExposeAsWebApp: 'true'
              ...
```

CORRECT VERSION (working):
```
captainVersion: 4
version: 3.3
services:
    '$$cap_appname':
        volumes:
            - /var/run/docker.sock:/var/run/docker.sock:ro
        caproverExtra:
            notExposeAsWebApp: 'true'
            dockerfileLines:
                - FROM mcuadros/ofelia:$$cap_version
                - CMD ["daemon", "--docker"]
              ...
```

I had to really dig to find an example from what was already published, which was time consuming to say the least.

After successful testing, I'll create a new fork of the one-click-apps, the added app, then perform a pull request.

**Comment by githubsaturn** [comment]
Thank you @maietta ! Looks like you're on the right track. I'll have limited access to internet for the next week or so. But I'll be able to help to get this up when I'm back.

**Comment by maietta** [comment]
> Thank you @maietta ! Looks like you're on the right track. I'll have limited access to internet for the next week or so. But I'll be able to help to get this up when I'm back.

You're welcome. Caprover is a very interesting project and enjoy using it for app deployments. It's well past time for me to make software contributions for it.

I've been forced to fork the Ofelia project, under a new name due to the 30 day wait on a fix already provided by someone else. Instead of waiting possibly forever, the new project name will is "Chronos".

I am currently setting up a new repository for "Chronos" at https://github.com/PremoWeb/Chronos and will be automating the creation of a docker image to be made available via Github's Container Registry, then updating the code for the One Click App to reflect this. After final testing, I'll go ahead and provide a Pull Request for the One Click App.

I very much appreciate your assistance. See ya on the other side.

**Comment by maietta** [comment]
I've submitted my first pull request to team Caprover as a relates to a new One Click App. I am the official source for the app, but couldn't find where to set this change to mark this as such.

See: https://github.com/caprover/one-click-apps/pull/507
